Anlysis of blob data set using KNN algorithm


In [23]:
#importing the pacakges required for Knn and generating blob sample data set 
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.datasets import make_blobs
import matplotlib.pyplot as plt
import numpy as np

centers = [[2, 4], [6, 6], [1, 9]]
n_classes = len(centers)
data, labels = make_blobs(n_samples=320, 
                          centers=np.array(centers),
                          random_state=1)

In [24]:
# Randomly generated data set, forming them into clusters and labling the data
print("Generated sample data", data)
print("Labling the data ",labels)

Generated sample data [[ 0.8808846   5.9560789 ]
 [ 2.31563495  1.97779878]
 [ 1.81343023  3.89825413]
 [-0.07296428  9.49515861]
 [ 0.5851531   9.45194604]
 [ 2.5627611   4.24073709]
 [ 4.69346593  6.07638048]
 [ 4.24407436  6.7147896 ]
 [ 2.50318481  5.29322588]
 [ 0.08420815  8.59274796]
 [ 4.18557541  2.60350366]
 [ 2.74055645  3.0462994 ]
 [ 1.82757179  3.12214158]
 [ 0.89465529  9.63019567]
 [ 2.86888616  4.75041164]
 [-0.63744959 12.9586027 ]
 [ 4.46120675  5.55210482]
 [ 5.52462712  6.47761018]
 [ 5.29865557  5.46277698]
 [ 6.1340482   7.20205486]
 [ 5.62519531  5.25003038]
 [ 1.68400133  8.64659002]
 [ 1.8858258   9.18986165]
 [-0.78791289  9.36184732]
 [ 1.65614632  4.04359686]
 [ 2.35054598  2.68771659]
 [-0.4614036   8.48365209]
 [ 4.52832571  3.75136522]
 [ 7.55880554  6.1094027 ]
 [ 1.59912181  4.82400562]
 [ 0.54005717  9.64435367]
 [ 6.10622272  4.47431968]
 [ 2.62091229  8.19381518]
 [ 2.14690038  8.95029742]
 [ 2.30017032  3.64775015]
 [ 1.87710977  3.06423057]
 [ 2.1

In [25]:
# split the data into training testing 
data_set = train_test_split(data, labels, 
                       train_size=0.8,
                       test_size=0.2,
                       random_state=12)
train_data, test_data, train_labels, test_labels = data_set

In [26]:
# Create and fit a nearest-neighbor classifier
from sklearn.neighbors import KNeighborsClassifier
# classifier "out of the box", no parameters
knn = KNeighborsClassifier()
knn.fit(train_data, train_labels)

KNeighborsClassifier()

In [27]:
# print some interested metrics
print("Predictions from the classifier:")
learn_data_predicted = knn.predict(train_data)
print(learn_data_predicted)
print("Target values:")
print(train_labels)
print(accuracy_score(learn_data_predicted, train_labels))

Predictions from the classifier:
[2 1 0 1 0 0 1 0 2 0 2 0 2 0 0 2 1 0 1 2 2 0 1 2 2 1 1 2 0 2 1 0 2 0 2 0 0
 0 1 2 1 2 0 1 2 2 2 0 0 2 2 1 1 1 0 2 2 1 2 2 0 1 2 0 0 1 0 2 1 0 0 2 2 0
 0 0 2 0 1 0 1 2 1 2 2 0 2 0 2 0 0 0 0 2 1 2 2 0 0 2 2 1 0 1 2 1 2 1 2 2 1
 1 0 1 1 1 1 0 0 2 1 1 1 1 2 2 1 0 2 0 2 2 2 0 2 0 1 0 1 1 2 1 0 1 2 1 2 2
 0 1 2 2 2 0 2 0 2 0 0 0 1 1 2 2 1 0 1 1 1 2 0 0 0 1 2 1 0 1 1 1 0 0 0 1 0
 1 0 1 1 1 2 2 2 0 0 0 2 2 2 1 2 2 0 1 2 1 1 2 0 2 2 1 1 0 2 2 1 1 0 2 1 0
 0 0 2 1 0 0 0 0 0 2 2 0 0 1 2 0 1 0 2 0 1 2 1 1 1 1 1 1 1 2 2 1 2 1]
Target values:
[2 1 0 1 0 0 1 0 2 0 2 0 2 0 0 2 1 0 1 2 2 0 1 1 2 1 1 2 0 2 1 0 2 0 2 0 0
 0 1 2 1 2 0 1 2 2 2 0 0 2 2 1 1 1 0 2 2 1 2 2 0 1 2 0 0 1 0 2 1 0 0 2 2 0
 0 0 2 0 1 0 1 2 1 2 2 0 2 0 2 0 0 0 0 2 1 2 2 0 0 2 2 1 0 1 2 1 2 1 2 2 1
 1 0 1 1 1 1 0 0 2 1 1 1 1 2 2 1 0 2 0 2 2 2 0 2 0 1 0 1 1 2 1 0 1 2 1 2 2
 0 1 2 2 2 1 2 0 2 0 0 0 1 1 2 2 1 0 1 1 1 2 0 0 0 1 2 1 0 1 1 1 1 0 0 1 0
 1 0 1 1 1 2 2 2 0 0 0 2 2 2 1 2 2 0 1 2 1 1 2 0 2 2 1 1 

In [28]:
# Applying knn classifier to test data
knn.fit(train_data, train_labels) 
test_data_predicted = knn.predict(test_data)
accuracy_score(test_data_predicted, test_labels)

0.984375

In [29]:
from sklearn.metrics import classification_report

print(classification_report(test_data_predicted, test_labels))

              precision    recall  f1-score   support

           0       1.00      0.96      0.98        26
           1       0.95      1.00      0.97        19
           2       1.00      1.00      1.00        19

    accuracy                           0.98        64
   macro avg       0.98      0.99      0.98        64
weighted avg       0.99      0.98      0.98        64



In [34]:
plt.scatter(data[0][:,0])
plt.scatter(data[0][:,0],data[0][:,1], c=data[1])

IndexError: ignored

As we know knn is an classification algorithm used to classify the data by traing the algorithm. With the pyhton packages we generated Artificial data set which is the combination of float,intigers. The data points are labled 0,1,2. The data is splitted into 80:20 as train and test.And applied Knn algortihm to the training data. In this data set there are no other parmaeters are involved in training data except the n_samples as 320. Then after traing the algorithm the accuracy of model on train data lables is 98%. Then to validate the model on test_data_predicted labels which also gives same accuracy. It states that model is working good. Here the accuracy of both sets are same. We can not say our model is very good. But we can say model is not overfitting because the values are quite comparable.The classification report also gives clear understanding over the perfomance of model.